The porpuse of this notebook is to clean data from null values, duplicates, structural errors, etc.

# Import

In [28]:
import pandas as pd
import numpy as np
import os

# Read data

In [3]:
FILE_PATH = os.path.join(os.pardir, 'data','raw','wines_SPA.csv')
data = pd.read_csv(FILE_PATH)
data.head(3)

winery           wine  year  rating  num_reviews country  \
0  Teso La Monja          Tinto  2013     4.9           58  Espana   
1         Artadi  Vina El Pison  2018     4.9           31  Espana   
2   Vega Sicilia          Unico  2009     4.8         1793  Espana   

             region   price                  type  body  acidity  
0              Toro  995.00              Toro Red   5.0      3.0  
1    Vino de Espana  313.50           Tempranillo   4.0      2.0  
2  Ribera del Duero  324.95  Ribera Del Duero Red   5.0      3.0

# Dimensions

In [4]:
data.shape

(7500, 11)

# Dtypes

In [5]:
data.dtypes

winery          object
wine            object
year            object
rating         float64
num_reviews      int64
country         object
region          object
price          float64
type            object
body           float64
acidity        float64
dtype: object

# Nulls

In [25]:
data.isnull().sum()

winery            0
wine              0
year              2
rating            0
num_reviews       0
country           0
region            0
price             0
type            545
body           1169
acidity        1169
dtype: int64

In [31]:
# Nulls in percentage
(data.isnull().sum() / len(data))*100

winery          0.000000
wine            0.000000
year            0.026667
rating          0.000000
num_reviews     0.000000
country         0.000000
region          0.000000
price           0.000000
type            7.266667
body           15.586667
acidity        15.586667
dtype: float64

# Q1: How resolve null values problem of "body", "acidity" and "type"?

## Are the null values of them the same index?

In [43]:
len(set(data[data['body'].isnull()].index) \
  & set(data[data['acidity'].isnull()].index))

1169

In [42]:
len(set(data[data['body'].isnull()].index) \
  & set(data[data['acidity'].isnull()].index)\
  & set(data[data['type'].isnull()].index))

545

It is verified that the three null parameters belong to the same indexes.

## Is it advisable to remove null data?

In [45]:
data.describe().round(2)

rating  num_reviews    price     body  acidity
count  7500.00      7500.00  7500.00  6331.00  6331.00
mean      4.25       451.11    60.10     4.16     2.95
std       0.12       723.00   150.36     0.58     0.25
min       4.20        25.00     4.99     2.00     1.00
25%       4.20       389.00    18.90     4.00     3.00
50%       4.20       404.00    28.53     4.00     3.00
75%       4.20       415.00    51.35     5.00     3.00
max       4.90     32624.00  3119.08     5.00     3.00

In [53]:
data[data['acidity'].notnull()].describe().round(2)

rating  num_reviews    price     body  acidity
count  6331.00      6331.00  6331.00  6331.00  6331.00
mean      4.26       444.14    65.71     4.16     2.95
std       0.12       734.10   162.64     0.58     0.25
min       4.20        25.00     4.99     2.00     1.00
25%       4.20       388.00    19.98     4.00     3.00
50%       4.20       402.00    29.15     4.00     3.00
75%       4.20       415.00    60.95     5.00     3.00
max       4.90     32624.00  3119.08     5.00     3.00

In [56]:
data[data['acidity'].notnull()].describe().round(2) - data.describe().round(3)

rating  num_reviews     price   body  acidity
count -1169.000    -1169.000 -1169.000  0.000    0.000
mean      0.005       -6.969     5.614  0.002    0.003
std       0.002       11.098    12.283 -0.003    0.002
min       0.000        0.000     0.000  0.000    0.000
25%       0.000       -1.000     1.080  0.000    0.000
50%       0.000       -2.000     0.620  0.000    0.000
75%       0.000        0.000     9.600  0.000    0.000
max       0.000        0.000     0.000  0.000    0.000

As we can see it is not advisable beacuse the difference in some characteristics would be high.

# Q1: Why is the year an object data type if it should be an integer?

In [6]:
data['year'].unique()

array(['2013', '2018', '2009', '1999', '1996', '1998', '2010', '1995',
       '2015', '2011', '2016', '1970', '1946', '1962', '2019', '2004',
       'N.V.', '1931', '1979', '2005', '2020', '2014', '1985', '1929',
       '2007', '2012', '2017', '2008', nan, '2006', '2000', '2003',
       '2002', '1991', '1994', '1990', '1989', '1987', '1986', '1981',
       '2001', '1968', '1964', '1982', '1974', '1983', '1955', '1980',
       '1972', '1953', '1958', '1942', '1965', '1992', '1973', '1997',
       '1967', '1975', '1910', '1961', '1954', '1988', '1969', '1951',
       '1928', '1976', '1949', '2021', '1959', '1922', '1978', '1925'],
      dtype=object)

It is observed that there are null values and the value "N.V.", which should be observed to which registers it belongs and evaluate some actions to be done.

# Q2: To which registry does "N.V" belongs?

In [7]:
data[data['year']=="N.V."]

winery                              wine  year  rating  \
20        Valdespino                  Toneles Moscatel  N.V.     4.8   
133       Barbadillo      Reliquia Palo Cortado Sherry  N.V.     4.7   
142           Alvear         Abuelo Diego Palo Cortado  N.V.     4.7   
143   Equipo Navazos             La Bota 78 de Oloroso  N.V.     4.7   
267          Osborne  Solera India Oloroso Rare Sherry  N.V.     4.6   
...              ...                               ...   ...     ...   
7388          Lustau         Candela Cream Dulce Sweet  N.V.     4.2   
7413          Lustau         Candela Cream Dulce Sweet  N.V.     4.2   
7438          Lustau         Candela Cream Dulce Sweet  N.V.     4.2   
7463          Lustau         Candela Cream Dulce Sweet  N.V.     4.2   
7488          Lustau         Candela Cream Dulce Sweet  N.V.     4.2   

      num_reviews country              region   price           type  body  \
20            174  Espana  Jerez-Xeres-Sherry  253.00         Sherry   4.0   
133            58  Espana  Jerez Palo Cortado  380.00         Sherry   4.0   
142            42  Espana    Montilla-Moriles  114.28  Pedro Ximenez   5.0   
143            41  Espana          Manzanilla   95.57         Sherry   4.0   
267            74  Espana  Jerez-Xeres-Sherry  189.99         Sherry   4.0   
...           ...     ...                 ...     ...            ...   ...   
7388          405  Espana  Jerez-Xeres-Sherry    7.10         Sherry   4.0   
7413          405  Espana  Jerez-Xeres-Sherry    7.10         Sherry   4.0   
7438          405  Espana  Jerez-Xeres-Sherry    7.10         Sherry   4.0   
7463          405  Espana  Jerez-Xeres-Sherry    7.10         Sherry   4.0   
7488          405  Espana  Jerez-Xeres-Sherry    7.10         Sherry   4.0   

      acidity  
20        3.0  
133       3.0  
142       1.0  
143       3.0  
267       3.0  
...       ...  
7388      3.0  
7413      3.0  
7438      3.0  
7463      3.0  
7488      3.0  

[288 rows x 11 columns]

In [8]:
(len(data[data['year']=="N.V."]) / len(data))*100 # 3.84%: The "N.V." values represent the 3.84% of the total.  

3.84

In [9]:
data['winery'].value_counts()

Contino                                 457
Artadi                                  261
La Rioja Alta                           254
Sierra Cantabria                        237
Matarromera                             232
                                       ... 
Briego                                    1
Guillem Carol - Cellers Carol Valles      1
Particular                                1
Bodegas Asenjo & Manso                    1
Joan Simo                                 1
Name: winery, Length: 480, dtype: int64

In [10]:
(data[data['year']=="N.V."]['winery'].value_counts() / len(data[data['year']=="N.V."]))*100 

Lustau                     77.430556
Bodegas Tradicion           1.736111
Osborne                     1.736111
Williams & Humbert          1.388889
Romate                      1.388889
Gonzalez-Byass              1.041667
Valdespino                  1.041667
Equipo Navazos              1.041667
Juan Pinero                 0.694444
Harveys                     0.694444
St. Petroni                 0.694444
Alvear                      0.694444
Altanza                     0.694444
El Maestro Sierra           0.694444
Alvaro Domecq               0.694444
Sacristia AB                0.694444
Hidalgo (La Gitana)         0.347222
Perez Barquero              0.347222
Santa Petronila             0.347222
Felix Solis                 0.347222
Bodegas Yzaguirre           0.347222
Bodegas Ximenez-Spinola     0.347222
Bodegas Arfe                0.347222
M. Antonio de la Riva       0.347222
Balandro                    0.347222
Tio Pepe                    0.347222
Vinos Guerra                0.347222
B

Lustau represents 77.43% of data where "year"=="N.V."

## Analyse Lustau winery

### Analyse wine in Lustau Winery

In [16]:
data[data['winery']=='Lustau']['wine'].value_counts()

Candela Cream Dulce Sweet                         220
Pedro Ximenez Murillo Seleccion del Centenario      1
Jerez-Xeres-Sherry 30 Year Old Oloroso VORS         1
Vermut Rojo                                         1
Name: wine, dtype: int64

### Analyse years in Lustau Winery

In [23]:
data[data['winery']=='Lustau']['year'].value_counts()

N.V.    223
Name: year, dtype: int64

100% of the year data in Lustau Winery is "N.V."

### Analyse region in Lustau Winery

In [18]:
data[data['winery']=='Lustau']['region'].value_counts()

Jerez-Xeres-Sherry    223
Name: region, dtype: int64

### Analyse type in Lustau Winery

In [19]:
data[data['winery']=='Lustau']['type'].value_counts()

Sherry           221
Pedro Ximenez      1
Name: type, dtype: int64

### Validate if it is possible to make an inference of the year from other parameters

In [24]:
data[['year','body','acidity']].corr()

/tmp/ipykernel_255/1822371770.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  data[['year','body','acidity']].corr()


body  acidity
body     1.00000 -0.01795
acidity -0.01795  1.00000